ydb/core/kqp/opt/logical/kqp_opt_log_indexes.cpp

```diff
@@ -334,15 +399,146 @@ TExprBase DoRewriteIndexRead(const TReadMatch& read, TExprContext& ctx,
+        .Columns(levelColumns)
+        .Settings(read.Settings())
+    .Done();
+
+    auto levelLambda = [&] {
+        const auto oldArgNodes = lambdaArgs.Children();
+        TNodeOnNodeOwnedMap replaces(oldArgNodes.size());
+        TExprNode::TListType newArgNodes;
+        newArgNodes.reserve(oldArgNodes.size());
+        for (const auto& arg : oldArgNodes) {
+            auto newArg = ctx.ShallowCopy(*arg);
+            YQL_ENSURE(replaces.emplace(arg.Get(), newArg).second);
+            newArgNodes.emplace_back(std::move(newArg));
+        }
+        return ctx.NewLambda(pos,
+            ctx.NewArguments(pos, std::move(newArgNodes)),
+            ctx.ReplaceNodes(TExprNode::TListType{lambdaBody.Ptr()}, replaces));
+    }();
+
+    auto topLevel0 = Build<TCoTop>(ctx, pos)
+        .Input(readLevel0)
+        // TODO(mbkkt) how to construct our own lambda?
+        //  Maybe good idea is construct lambda with knn udf and two member access as arguments
+        //   and then replace one of them to argument without access to indexed field...
+        .KeySelectorLambda(levelLambda)
+        .SortDirections(top.SortDirections())
+        .Count(count)
+    .Done();
+
+    auto mapLevel0 = Build<TCoMap>(ctx, pos)
+        .Input(topLevel0)
+        .Lambda()
+            .Args({mapArg})
+            .Body<TCoAsStruct>().Add(mapMembers).Build()
+        .Build()
+    .Done();
+
+    auto postingRead = Build<TKqlLookupTable>(ctx, pos)
+        .Table(postingTable)
+        .LookupKeys(mapLevel0)
+        .Columns(postingColumns)
+    .Done();
+
+    auto mainRead = Build<TKqlLookupTable>(ctx, pos)
+        .Table(mainTable)
+        .LookupKeys(postingRead)
+        .Columns(mainColumns)
+    .Done().Ptr();
+
+    if (flatMap) {
+        mainRead = Build<TCoFlatMap>(ctx, flatMap.Cast().Pos())
+            .Input(mainRead)
+            .Lambda(ctx.DeepCopyLambda(flatMap.Cast().Lambda().Ref()))
+        .Done().Ptr();
+    }
+
+    auto mainTop = Build<TCoTopBase>(ctx, top.Pos())
+        .CallableName(top.Ref().Content())
+        .Input(mainRead)
+        .KeySelectorLambda(ctx.DeepCopyLambda(top.KeySelectorLambda().Ref()))
+        .SortDirections(top.SortDirections())
+        .Count(top.Count())
+    .Done();
+
+    return mainTop;
+}
+
 } // namespace
 
 TExprBase KqpRewriteIndexRead(const TExprBase& node, TExprContext& ctx, const TKqpOptimizeContext& kqpCtx) {
     if (auto indexRead = TReadMatch::Match(node)) {
 
         const auto& tableDesc = GetTableData(*kqpCtx.Tables, kqpCtx.Cluster, indexRead.Table().Path());
-        const auto& [indexMeta, _ ] = tableDesc.Metadata->GetIndexMetadata(TString(indexRead.Index().Value()));
+        const auto indexName = indexRead.Index().Value();
+        auto [implTable, indexDesc] = tableDesc.Metadata->GetIndex(indexName);
+        // TODO(mbkkt) instead of ensure should be warning and main table read?
+        YQL_ENSURE(indexDesc->Type != TIndexDescription::EType::GlobalSyncVectorKMeansTree,
```

Comment:
maybe move this logic and ensure to the function?

Reply:
I don't really understand where you suggest to move this function, can you point me?